### PR TITLE
Move Processing to Hooks and Filters REDUX

### PR DESCRIPTION
**Pull request by dshanske**
Allows removing/replacing of upload handling, geodata saving, and mf2 storing to replace with custom functions as needed.

The difference between this and #26 is that this switches it from post-processing to pre-processing. The advantage to this is that as opposed to just removing the data...still an option...it can be transformed by a filter later in the process.

### CODE REVIEW

File: micropub.php [hunk 1]
```diff
@@ -61,7 +61,16 @@ public static function init() {
     add_action('send_headers', array('Micropub', 'http_header'));
     add_filter('host_meta', array('Micropub', 'jrd_links'));
     add_filter('webfinger_data', array('Micropub', 'jrd_links'));
-  }
+
+    // Store MF2
+    add_filter('before_micropub', array('Micropub', 'store_mf2'), 8);
```
Comment (snarfed) [inline micropub.php]: is the priority 8 here important?

Comment (dshanske) [inline micropub.php]: Default priority is 10. Storing the MF2 in the meta should occur before the others, so I set it to 8. That means other filters will have access to it in the args array.

File: micropub.php [hunk 2]
```diff
@@ -310,18 +316,14 @@ private static function generate_args() {
...
     }
-    // If the theme declares it supports microformats2, pass the content through
-    if (current_theme_supports('microformats2')) {
-      if (isset($_POST['content'])) {
-        $args['post_content'] = $_POST['content'];
-      }
-      else if (isset($_POST['summary'])) {
-        $args['post_content'] = $_POST['summary'];
-      }
+    if (!isset($args['meta_input'])) {
```
Comment (snarfed) [inline micropub.php]: `meta_input` isn't used in this method, so i think you can remove this block.

File: micropub.php [hunk 3]
```diff
@@ -432,18 +452,21 @@ private static function postprocess($post_id) {
       $geo = explode(':', $geo);
       $geo = explode(';', $geo[0]);
       $coords = explode(',', $geo[0]);
-      update_post_meta($post_id, 'geo_latitude', trim($coords[0]));
-      update_post_meta($post_id, 'geo_longitude', trim($coords[1]));
+      $args['meta_input']['geo_latitude'] = trim($coords[0]);
```
Comment (snarfed) [inline micropub.php]: if we want to delay storing the post meta, that's ok, but we do still need to store it at some point. it looks like you removed the `updated_post_meta` alls here but didn't add them anywhere else.

Comment (snarfed) [inline micropub.php]: you might consider adding a few unit tests in this PR, e.g. checking that post meta is stored. :P

File: micropub.php [hunk 4]
```diff
@@ -452,14 +475,14 @@ private static function store_mf2($post_id) {
       }
       if (!in_array($key, $blacklist)) {
         $key = 'mf2_' . $key;
-        delete_post_meta($post_id, $key);  // clear old value(s)
         foreach ($value as $val) {
           if (!empty($val)) {
-            add_post_meta($post_id, $key, $val);
+            $args['meta_input'][$key] = $val;
```
Comment (snarfed) [inline micropub.php]: ditto the removed `delete_post_meta` and `add_post_meta` calls here.

File: micropub.php [hunk 5]
```diff
@@ -330,7 +332,15 @@ private static function generate_args() {
    * Generates and returns a post_content string suitable for wp_insert_post()
    * and friends.
    */
-  private static function generate_post_content() {
+  public static function generate_post_content($args) {
+    // If the theme declares it supports microformats2, pass the content through
+    if (current_theme_supports('microformats2')) {
+      return $args;
+		}
+    // Disable if Kind Taxonomy is enabled
```
Comment (snarfed) [inline micropub.php]: i wonder if there's a way to do this without special casing the post kinds plugin explicitly like this?

File: micropub.php (hunk 1 above)
Comment (snarfed) [inline micropub.php]: sgtm. probably worth adding a comment.

File: micropub.php (hunk 2 above)
Comment (dshanske) [inline micropub.php]: I wanted it created by default, but point taken.

File: micropub.php (hunk 5 above)
Comment (dshanske) [inline micropub.php]: I could move that code back to the Post Kinds plugin. I had it there originally.

File: micropub.php (hunk 3 above)
Comment (dshanske) [inline micropub.php]: wp_insert_post, which is used in new and updated post entries, takes anything stored in meta_input and updates the  post meta already. That is why I switched it. It has advantages and removes some code.

Also, checked...it is being stored.

File: micropub.php (hunk 4 above)
Comment (dshanske) [inline micropub.php]: deleting and then adding isn't necessary. See above.

**Comment by snarfed** [comment]
if you're bumping the version now, mind starting a version 0.5 section in readme.txt?

File: micropub.php (hunk 3 above)
Comment (snarfed) [inline micropub.php]: > The feature wasn't added till 4.4 though, which bumps the minimum version number to 4.4. But no one should be running a version older than 2 releases back anyway and be supported.

tough talk! :P

right now we only require >=3.0.1: https://wordpress.org/plugins/micropub/ . bumping that up to 4.4 is a big jump, but i'm ok with it if you think we should.

Comment (snarfed) [inline micropub.php]: > Also, checked...it is being stored.

manual checks are fleeting. unit tests are forever. :P

File: micropub.php (hunk 5 above)
Comment (snarfed) [inline micropub.php]: got it. i don't feel strongly, and you know the wordpress conventions better than me. whatever you think is best!

**Comment by dshanske** [comment]
I agree it is a tough stance. I will try unit tests, still bad at it. Will add a readme. Also, as for conventions, two versions back is what most projects guarantee for Core.

**Comment by snarfed** [comment]
> two versions back is what most projects guarantee for Core.

aha! so then 4.4 is maybe a bit too aggressive, since it's only one version back.

**Comment by dshanske** [comment]
4.6 will be out shortly.

**Comment by dshanske** [comment]
I added a unit test.

File: micropub.php [hunk 6]
```diff
@@ -337,7 +337,8 @@ public static function generate_post_content($args) {
     if (current_theme_supports('microformats2')) {
       return $args;
 		}
-    // Disable if Kind Taxonomy is enabled
+    // Disable if Kind Taxonomy is enabled. The Kind Taxonomy is used in the Post Kinds plugin.
+    // The Post Kinds plugin handles its own markup.
```
Comment (snarfed) [inline micropub.php]: the valuable part to add here is the link, eg https://wordpress.org/plugins/indieweb-post-kinds/

File: micropub.php [hunk 7]
```diff
@@ -62,7 +62,7 @@ public static function init() {
     add_filter('host_meta', array('Micropub', 'jrd_links'));
     add_filter('webfinger_data', array('Micropub', 'jrd_links'));
 
-    // Store MF2
+    // Store MF2 - set at priority 8 in order to occur before the other filters.
```
Comment (snarfed) [inline micropub.php]: the valuable thing to comment here is *why* we want it to occur before the other filters. :P

**Comment by snarfed** [comment]
awesome! nice work on the test! merging now. i'm going to convert this to wordpress style, so i'll also tweak the comments in that commit.

congrats!

**Pull request merged**